### PR TITLE
feat(utilities): add `enumToObject`

### LINKED ISSUE #507: request: `@sapphire/bitfield` lacks support for 2-way enums
### Is there an existing issue or pull request for this?

- [X] I have searched the existing issues and pull requests

### Feature description

Depends on #502.

Implementing a `mapEnum` utility function to map TypeScript's 2-way enums into 1-way ones, as `BitField`'s constructor requires all values to be of the same type, and supporting them in the constructor can lead to much more complicated types, as opposed to a function, which would remove one step.

### Desired solution

The idea is to make a function that maps (both value and types) of 2-way enums (aka TypeScript enums) into 1-way ones, for example:

```typescript
enum State {
	On,
	Off
}
```

Results on the following:

```typescript
const State = {
	On: 0,
	Off: 1,
	0: 'On',
	1: 'Off'
};
```

The idea is to have a function, e.g. `mapEnum`, that maps the above object (and type) into:

```typescript
const State = {
	On: 0,
	Off: 1
};

type State = {
	On: 0;
	Off: 1;
};
```

So it can be passed into the BitField's constructor in the following manner:

```typescript
const PermissionsBitField = new BitField(mapEnum(PermissionsFlagsBits));
```

### Alternatives considered

Mapping in the constructor by ignoring values which type is `string` and are a property of `flags`, but then again types would need to handle this, and I fear they might be incredibly complex.

### Additional context

Conversation at https://github.com/sapphiredev/utilities/pull/502#discussion_r996082235

### PR DESCRIPTION
**Pull request by kyranet**
Fixes #507

### CODE REVIEW

File: packages/bitfield/src/lib/utilities.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * // 	1: 'Read',
+ * // 	2: 'Write'
+ * // }
+ *
+ * enumToObject(Permissions);
+ * // { Read: 1, Write: 2 }
+ * ```
+ * @returns The mapped object
+ */
+export function enumToObject<T extends object>(enumObject: T): { [K in Exclude<keyof T, `${number}`>]: T[K] } {
```
Comment (vladfrangu) [inline packages/bitfield/src/lib/utilities.ts:22]: Doesn't this allow passing in null too?

Comment (kyranet) [reply packages/bitfield/src/lib/utilities.ts:22]: No, we removed `NonNullObject` because of this.

File: packages/bitfield/src/lib/utilities.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	const result = {} as any;
+	for (const [key, value] of Object.entries(enumObject)) {
+		if (Number.isNaN(Number(key))) result[key] = value;
+	}
+
+	return result as any;
```
Comment (favna) [inline packages/bitfield/src/lib/utilities.ts:28]: any? Use Generic type.

**Review by favna — CHANGES_REQUESTED** [review]

File: packages/bitfield/src/lib/utilities.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * // 	2: 'Write'
+ * // }
+ *
+ * enumToObject(Permissions);
+ * // { Read: 1, Write: 2 }
+ * ```
+ * @returns The mapped object
+ */
+export function enumToObject<T extends object>(enumObject: T): { [K in Exclude<keyof T, `${number}`>]: T[K] } {
+	const result = {} as any;
```
Comment (favna) [inline packages/bitfield/src/lib/utilities.ts:23]: There's still an any here...

**Review by favna — CHANGES_REQUESTED** [review]

**Review by favna — APPROVED** [review]
Loads better. Now if anything ever changes in TS that would affect this we'll actually know.

**Pull request merged**